### PR TITLE
Isaicp 4645: Sort out duplicate entities in CTT

### PR DESCRIPTION
**Pull request by idimopoulos**

### CODE REVIEW

File: web/modules/custom/joinup_core/joinup_core.post_update.php [hunk 1]
```diff
@@ -90,3 +94,272 @@ function joinup_core_post_update_remove_action_transfer_solution_ownership() {
...
+
+    foreach ($duplicate_ids as $duplicate_id) {
+      if (!($duplicate_group = Rdf::load($duplicate_id))) {
+        continue;
+      }
+      break;
+    }
+
+    if (empty($duplicate_group)) {
+      return;
```
Comment (claudiu-cristea) [inline web/modules/custom/joinup_core/joinup_core.post_update.php]: I guess this is `continue;`, not `return;`.

File: web/modules/custom/joinup_core/joinup_core.post_update.php [hunk 2]
```diff
@@ -90,3 +94,272 @@ function joinup_core_post_update_remove_action_transfer_solution_ownership() {
...
+      $content_ids = $membership_manager->getGroupContentIds($duplicate_group);
+      foreach ($content_ids as $entity_type => $entity_ids) {
+        // Avoid processing other group content like og menu instances.
+        if (!in_array($entity_type, ['node', 'rdf_entity'])) {
+          continue;
+        }
+        $storage = \Drupal::entityTypeManager()->getStorage($entity_type);
+        foreach ($storage->loadMultiple($entity_ids) as $entity) {
+          $entity->set(OgGroupAudienceHelperInterface::DEFAULT_FIELD, $original_id);
+          $entity->skip_notification = 1;
```
Comment (claudiu-cristea) [inline web/modules/custom/joinup_core/joinup_core.post_update.php]: s/1/TRUE

File: web/modules/custom/joinup_core/joinup_core.post_update.php [hunk 3]
```diff
@@ -90,3 +94,272 @@ function joinup_core_post_update_remove_action_transfer_solution_ownership() {
...
+  }
+}
+
+/**
+ * Remove ctt duplicates.
+ */
+function _joinup_core_post_update_ctt_duplicates_delete_duplicates() {
+  foreach (_joinup_core_get_duplicated_ids() as $original_id => $duplicated_ids) {
+    foreach (Rdf::loadMultiple($duplicated_ids) as $entity) {
+      $entity->skip_notification = 1;
```
Comment (claudiu-cristea) [inline web/modules/custom/joinup_core/joinup_core.post_update.php]: s/1/TRUE

File: web/modules/custom/joinup_core/joinup_core.post_update.php [hunk 4]
```diff
@@ -90,3 +94,272 @@ function joinup_core_post_update_remove_action_transfer_solution_ownership() {
...
+      }
+
+      // Unset the relationship to any distribution so that they can properly
+      // moved.
+      $duplicate_group->set('field_is_distribution', NULL);
+      $duplicate_group->skip_notification = 1;
+      $duplicate_group->save();
+
+      $content_ids = $membership_manager->getGroupContentIds($duplicate_group);
+      foreach ($content_ids as $entity_type => $entity_ids) {
```
Comment (claudiu-cristea) [inline web/modules/custom/joinup_core/joinup_core.post_update.php]: Let's use the good naming practices from the core `$entity_type_id` (`$entity_type` refers to the object).

File: web/modules/custom/joinup_core/joinup_core.post_update.php [hunk 5]
```diff
@@ -90,3 +94,272 @@ function joinup_core_post_update_remove_action_transfer_solution_ownership() {
...
+ */
+function _joinup_core_post_update_ctt_duplicates_collection_relations() {
+  // The id of the ctt collection.
+  $id = 'http://administracionelectronica.gob.es/ctt';
+
+  $collection = Rdf::load($id);
+  foreach (_joinup_core_get_duplicated_ids() as $original_id => $duplicate_ids) {
+    $collection->field_ar_affiliates->appendItem($original_id);
+  }
+  $collection->skip_notification = 1;
```
Comment (claudiu-cristea) [inline web/modules/custom/joinup_core/joinup_core.post_update.php]: s/1/TRUE

File: web/modules/custom/joinup_core/joinup_core.post_update.php [hunk 6]
```diff
@@ -90,3 +94,272 @@ function joinup_core_post_update_remove_action_transfer_solution_ownership() {
...
+      foreach ($duplicate_memberships as $key => $duplicate_membership) {
+        // If the membership does not exist in the original entity, transfer it
+        // and maintain the values.
+        if (!isset($memberships[$key])) {
+          $duplicate_membership->setGroup($group);
+          $duplicate_membership->skip_notification = 1;
+          $duplicate_membership->save();
+        }
+        // If the membership does exist in the original entity, add the roles to
+        // the existing membership. We do not need to filter anything out since
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/joinup_core/joinup_core.post_update.php]: You mean array_diff them, right?

[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/joinup_core/joinup_core.post_update.php]: Yeah, diff. The idea is to not perform an unneeded  save.

**Comment by idimopoulos** [comment]
> This one I didn't researched if is possible: I see that each helper is iterating over _joinup_core_get_duplicated_ids(). Is it not possible, theoretically, to do a single iteration in the main function and then call the helpers inside this iteration to do their job? Just askin'
It would probably work well too (haven't tried it). It would be either of both. The entities loaded are cached anyway and all actions are taking place in a single batch process. 
In the end it all comes down to how you design it in your mind. Either group actions together, or group entities together. Since I had bullet points as tasks, I simply followed the logic so that also the commits are clearer.

**Pull request merged**